modules/videoio/test/test_gstreamer.cpp

```diff
@@ -0,0 +1,259 @@
+{
+typedef cv::Point_<double> Point2d;
+typedef tuple<string, string> Param;
+
+class Videoio_Gstreamer_Test : public testing::TestWithParam<Param>
+{
+protected:
+    Videoio_Gstreamer_Test()
+    {
+        format = get<0>(GetParam());
+        pattern = get<1>(GetParam());
+        size = Size(640, 480);
+        size_nv12 = Size(640, 720);
+        count_frames = 10;
+        pipeline << "videotestsrc pattern=" << pattern;
+        pipeline << " num-buffers=" << count_frames;
+        if (format == "video/x-bayer")
+        {
+            pipeline << " ! " << format;
+        }
+        else if (format == "image/jpeg")
+        {
+            pipeline << " ! jpegenc ! " << format;
+        }
+        else
+        {
+            pipeline << " ! video/x-raw, format=" << format;
+        }
+        pipeline << ", width=" << size.width << ", height=" << size.height;
+        pipeline << " ! appsink sync=false";
+    }
+
+    void testSizeFormats()
+    {
+        VideoCapture cap(pipeline.str(), CAP_GSTREAMER);
+        ASSERT_TRUE(cap.isOpened());
+        for (int i = 0; i < count_frames; ++i)
+        {
+            Mat frame, gray_frame;
+            cap >> frame;
+            if (frame.empty()) { continue; }
+            if (format == "image/jpeg")
+            {
+                Mat buffer = imdecode(frame, cv::IMREAD_UNCHANGED);
+                frame = buffer;
+            }
+
+            if (frame.empty()) { continue; }
+            if ((format == "NV12") || (format == "NV21") ||
+                (format == "YV12") || (format == "I420") )
+            {
+                EXPECT_EQ(size_nv12, frame.size());
+            }
+            else
+            {
+                EXPECT_EQ(size, frame.size());
+            }
+
+            double actual = cap.get(CAP_PROP_POS_FRAMES);
+            if (i + 1 != (int)actual)
+            {
+              std::cout << "Frame " << i << " index is " << actual << std::endl;
+            }
+            EXPECT_EQ(i + 1, static_cast<int>(actual));
+        }
+        {
+            Mat frame;
+            cap >> frame;
+            EXPECT_TRUE(frame.empty());
+        }
+        cap.release();
+        ASSERT_FALSE(cap.isOpened());
+    }
+
+    void testObjectStructure()
+    {
+        VideoCapture cap(pipeline.str(), CAP_GSTREAMER);
+        ASSERT_TRUE(cap.isOpened());
+
+        double sum_vertical_diameter   = 0.0;
+        double sum_horizontal_diameter = 0.0;
+        for (int i = 0; i < count_frames; ++i)
+        {
+            Mat frame, gray_frame;
+            if (cap.grab()) { cap.retrieve(frame); } else { continue; }
+            if (format == "image/jpeg")
+            {
+                Mat buffer = imdecode(frame, cv::IMREAD_UNCHANGED);
+                cvtColor(buffer, frame, CV_RGB2GRAY);
+            }
+            gray_frame = getGrayFrame(frame);
+
+            std::vector<Point2d> vertical, horizontal;
+            vertical = getDiameterLine("vertical", gray_frame);
```

Comment:
Check reference functions implemented in `cvtest` namespace [here](https://github.com/opencv/opencv/blob/099a16bd86677b807d5d8fe055b06cacb8d5f462/modules/ts/include/opencv2/ts.hpp#L197-L252), maybe some of them could help to simplify the frame content verification?